python/pyarrow/dataset.py

```diff
@@ -733,19 +733,17 @@ def write_dataset(data, base_dir, basename_template=None, format=None,
     """
     from pyarrow.fs import _resolve_filesystem_and_path
 
-    if isinstance(data, Dataset):
-        schema = schema or data.schema
```

Comment:
We are extracting the `schema` here, but this is not actually used any more, if the user passes a Dataset or Scanner. 

This should be better documented, but to avoid surprises, could also error in that case? (or only if the schema's are not equal)

Reply:
Hmm, would the user ever want to pass a schema to force a cast? Though I suppose that's redundant (either configure the scanner or pass a scanner instead of a dataset).

I think a deprecation warning would be appropriate here (schema will only be allowed when passing an iterable in the future).

Reply:
> Hmm, would the user ever want to pass a schema to force a cast? Though I suppose that's redundant (either configure the scanner or pass a scanner instead of a dataset).

In that case, I think the user could have created the dataset with a specific (forced) schema: `ds.dataset(..., schema=schema)`. 
But once you have a dataset, it's not easy to change the schema. The `scanner` method doesn't take a `schema` keyword (although that's maybe something to add, to project to the specified schema). 

Yes, if we would want to raise, we can indeed raise a warning first.

Reply:
Yeah, either accepting a schema for a projection or having a separate schema keyword would be a nice convenience if all you want to do is select/cast columns.